Removes extra count in Packages list after a search in PM UI

dominoFire opened the pull request:
## Bug

Fixes:

- https://github.com/NuGet/Home/issues/9077
- https://devdiv.visualstudio.com/DevDiv/_workitems/edit/994446

 
Regression: No
* Last working version: N/A
* How are we preventing it in future: Use standard WPF/VS controls, if possible

## Fix

Details: 
- Adds the "Loading...", just before the search results items are loaded
- Removes the "Loading..." indicator, always
- Adds some code guards in existing methods
- Adds a string resource to help narrator identify the ListBox inside the InfinteScrollList control
- Adds documentation

## Testing/Validation

Tests Added: Yes
Reason for not adding tests: UI features, 
Validation:  
## Accessibility Insights run

Found an issue with Accessibility Insights. See images below. 

### Test

![image](https://user-images.githubusercontent.com/1192347/75080201-35b91f00-54c0-11ea-8aae-90e6b9d43fd2.png)


### Before

![image](https://user-images.githubusercontent.com/1192347/75078421-69914600-54ba-11ea-83c3-242f41f26236.png)


### After

![image](https://user-images.githubusercontent.com/1192347/75079869-2685a180-54bf-11ea-9ddf-78bc782f2c05.png)



## Manual validation

Small live demo with a slow network simulation, thanks to https://jagt.github.io/clumsy/

Commit:
First draft

Commit:
Updated documentation

Commit:
Min-changeset

Commit:
Typo

Commit:
Added AutomationProperties.Name to ListBox in InfiniteScrollList

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollList.xaml.cs:
```diff
@@ -37,7 +37,9 @@ public partial class InfiniteScrollList : UserControl
 
-        // This exists only to facilitate unit testing.
+        /// <summary>
+        /// This exists only to facilitate unit testing. It is triggered after LoadItems() is just before finished
```
It is triggered after LoadItems(), just before it is finished.

dominoFire replied:
Rephrased.

rrelyea commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollList.xaml.cs:
```diff
@@ -177,11 +183,17 @@ private void LoadItems(PackageItemListViewModel selectedPackageItem, Cancellatio
             var loadCts = CancellationTokenSource.CreateLinkedTokenSource(token);
             Interlocked.Exchange(ref _loadCts, loadCts)?.Cancel();
 
+            // add Loading... indicator if not present
```
comment doesn't match code.

dominoFire replied:
Changed to correct meaning: add the loading indicator at the beginning.

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollList.xaml.cs:
```diff
@@ -235,6 +247,7 @@ private void LoadItems(PackageItemListViewModel selectedPackageItem, Cancellatio
                 UpdateCheckBoxStatus();
 
                 LoadItemsCompleted?.Invoke(this, EventArgs.Empty);
+                Items.Remove(_loadingStatusIndicator);
```
Wrap this in a finally {} block from the outermost try/catch ?
We call SetError() on the _loadingStatusBar, but does that hide the Indicator or does it stay displayed infinitely?

donnie-msft replied:
If SetError already hides the indicator, then this line should be moved inside the outermost try {} logic so that it only removes the item if no Exception occurred.

dominoFire replied:
Nice catch!

> Wrap this in a finally {} block from the outermost try/catch ?

The idea is to add the indicator at the very beginning and remove it at the very end of the method execution. I prefer to leave it outside of the try cath

> We call SetError() on the _loadingStatusBar, but does that hide the Indicator or does it stay displayed infinitely?

If we look futher, `_loadingStatusBar.SetError()` only sets an internal error message. I didn't find any bindings or references of the error message outside of the `LoadingStatusIndicator` class. So, I will remove that property.

> If SetError already hides the indicator, then this line should be moved inside the outermost try {} logic so that it only removes the item if no Exception occurred.

For the reasons stated above, I prefer to leave it outside.

This lead me to discover an issue with this PR. See comments added.

dominoFire replied:
Nice catch!

> Wrap this in a finally {} block from the outermost try/catch ?

The idea is to add the indicator at the very beginning and remove it at the very end of the method execution. I prefer to leave it outside of the try cath

> We call SetError() on the _loadingStatusBar, but does that hide the Indicator or does it stay displayed infinitely?

If we look futher, `_loadingStatusBar.SetError()` only sets an internal error message. I didn't find any bindings or references of the error message outside of the `LoadingStatusIndicator` class. So, I will remove that property.

> If SetError already hides the indicator, then this line should be moved inside the outermost try {} logic so that it only removes the item if no Exception occurred.

For the reasons stated above, I prefer to leave it outside.

This lead me to discover an issue with this PR. See comments added.

dominoFire replied:
I managed to solve the issue documented in the previous TODO. At the end, I moved the loading indicator removal to a `finally` block

rrelyea commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollList.xaml.cs:
```diff
@@ -399,6 +417,11 @@ private Visibility EvaluateStatusBarVisibility(IItemLoader<PackageItemListViewMo
+        /// <summary>
+        /// Appends <c>packages</c> to the <c>Items</c>
+        /// </summary>
+        /// <param name="packages">View model collection to add</param>
```
terminology is weird...  you are adding all the items from the view model collection to the items collection.
can you make clearer?

dominoFire replied:
Rephrased the docs for this method.

rrelyea commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollList.xaml.cs:
```diff
@@ -407,7 +430,7 @@ private void UpdatePackageList(IEnumerable<PackageItemListViewModel> packages, b
                 {
                     // remove the loading status indicator if it's in the list
-                    Items.Remove(_loadingStatusIndicator);
+                    var removed = Items.Remove(_loadingStatusIndicator);
```
don't use var here?

dominoFire replied:
Changed to explicit bool type declaration.

Commit:
Addressed Feedback #1

- Explicit var
- Condition to leave status indicator in case of error

Commit:
Feedback 2

- Fixed dissapearing 'No items found'
- Removed unused ErrorMessage in LoadingStatusIndicator

dominoFire committed:
Fix 'No items found' message

dominoFire committed:
Added finally block. Typo

dominoFire committed:
Added test

rrelyea commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollList.xaml.cs:
```diff
@@ -226,11 +240,24 @@ private void LoadItems(PackageItemListViewModel selectedPackageItem, Cancellatio
                 }
+                finally
+                {
+                    if (_loadingStatusIndicator.Status != LoadingStatus.NoItemsFound)
```
i thought when we talked on friday...we wanted to ensure that we don't remove the statusindicator if either of those 2 exceptions happen. your current check checks for "empty" list...does it handle those exceptions well. what about for multi-feed searches?

dominoFire replied:
I will revert the error messages in the exceptions.

dominoFire replied:
Multi-feed search can fail and lead to an error. Documented in this issue:   https://github.com/NuGet/Home/issues/9218

rrelyea reviewed (commented):
ok with everything except my question about the finally block in the error cases.

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollList.xaml.cs:
```diff
@@ -158,6 +161,10 @@ public bool CheckBoxesEnabled
+        /// <summary>
+        /// Keep the selected package on the item after a search
+        /// or select the first on the search if none was selected before
+        /// <param name="selectedItem">Previously selected item</param>
```
first line, add period: "before."
second line: add something like " will be selected, if found. Otherwise, select the first package"

zivkan replied:
the `summary` xml tag is also not closed.

dominoFire replied:
- Fixed `summary` tag.
- Rephrased

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollList.xaml.cs:
```diff
@@ -183,8 +190,17 @@ private void LoadItems(PackageItemListViewModel selectedPackageItem, Cancellatio
+
                 try
                 {
+                    // add Loading... indicator if not present
```
Consider making this a helper method? eg, `bool AddSingleLoadingIndicator(List<stuff> Items)`

dominoFire replied:
Generalizing this part is tricky, because in each part in which we add the status indicator, we need different pieces of information.

dominoFire replied:
Following up in another issue: https://github.com/NuGet/Home/issues/9219

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollList.xaml.cs:
```diff
@@ -226,11 +240,24 @@ private void LoadItems(PackageItemListViewModel selectedPackageItem, Cancellatio
+                {
+                    if (_loadingStatusIndicator.Status != LoadingStatus.NoItemsFound)
+                    {
+                        var emptyListCount = addedLoadingIndicator ? 1 : 0;
```
Another good Helper Method candidate...

dominoFire replied:
See https://github.com/NuGet/NuGet.Client/pull/3254#discussion_r384018522

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollList.xaml.cs:
```diff
@@ -226,11 +240,24 @@ private void LoadItems(PackageItemListViewModel selectedPackageItem, Cancellatio
                 }
+                finally
+                {
+                    if (_loadingStatusIndicator.Status != LoadingStatus.NoItemsFound)
```
Think the entire Finally should be a Helper Method. eg, `void RemoveLoadingIndicatorIfPresent(List<stuff> Items)`

dominoFire replied:
We only have this case in this finally block. That would lead to a helper method with only one usage.

dominoFire replied:
Following up in another issue. https://github.com/NuGet/Home/issues/9219

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.UI.Test/Xamls/InfiniteScrollListTests.cs:
```diff
@@ -269,5 +279,197 @@ public async Task LoadItems_BeforeGettingCurrent_WaitsForInitialResults()
+        }
+
+        /// <summary>
+        /// Test implementation of <see cref="IPackageSearchMetadata"/>. Not for produciton use
```
typo:production

dominoFire replied:
Fixed typo. Moved class to a new file.

Commit:
Feedback 3

- Separated test classes
- Typos
- Reverted Error Message

rrelyea commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollList.xaml.cs:
```diff
@@ -231,6 +249,22 @@ private void LoadItems(PackageItemListViewModel selectedPackageItem, Cancellatio
                 }
+                finally
+                {
+                    if (_loadingStatusIndicator.Status != LoadingStatus.NoItemsFound
```
if logic needs to be improved. think you want this to be an &&

dominoFire replied:
Yes. I want to execute the the code block if and only if status is not `NoItemsFound` neither `ErrorOccurred`

rrelyea reviewed (approved):
once you fix that, all set from me.

Commit:
Fixed condition. Added notes
